apps/api/src/api/states.ts

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2024 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { ApplicationStates, ApplicationStateValues } from '@pcgl-daco/data-model/src/types.js';
+import { ITransition, StateMachine, t as transition } from 'typescript-fsm';
+import { getDbInstance } from '../db/index.js';
+import { applications } from '../db/schemas/applications.js';
+import applicationService from '../service/application-service.js';
+import { AsyncResult, failure, success } from '../utils/results.js';
+import { validateContent } from './validation.js';
+
+const {
+	DRAFT,
+	INSTITUTIONAL_REP_REVIEW,
+	REP_REVISION,
+	DAC_REVIEW,
+	DAC_REVISIONS_REQUESTED,
+	REJECTED,
+	APPROVED,
+	CLOSED,
+	REVOKED,
+} = ApplicationStates;
+
+export enum ApplicationStateEvents {
+	submit = 'submit',
+	close = 'close',
+	edit = 'edit',
+	revision_request = 'revision_request',
+	approve = 'approve',
+	reject = 'reject',
+	revoked = 'revoked',
+}
+
+const { submit, close, edit, revision_request, approve, reject, revoked } = ApplicationStateEvents;
+
+// TODO: Replace with individual methods
+type ApplicationTransitionCallback = () => AsyncResult<string>;
+
+type ApplicationTransitions = ITransition<
+	ApplicationStateValues,
+	ApplicationStateEvents,
+	ApplicationTransitionCallback
+>;
+
+export class ApplicationStateManager extends StateMachine<ApplicationStateValues, ApplicationStateEvents> {
+	private readonly _id: number;
+	private readonly _application: typeof applications.$inferSelect;
+	private readonly _initState: ApplicationStateValues;
+
+	// Handler Methods
+	// Submit
+	async submitDraft() {
+		if (this.can(submit)) {
+			// TODO: Add Validation
+			const validationResult = await validateContent(this._application);
+			if (validationResult.success) {
+				await this.dispatch(submit);
+				return validationResult;
```

Comment:
Example public method. Validates state transition and related record before allowing event.

Reply:
Looking at organization, I'm wondering if I should move private methods to the top (because they're used in the constructor) then public methods to the bottom (convention to have the constructor at the top, public methods used after instantiation, etc)